[10214] Release Twisted 21.7.0 (#1614)
twisted/twisted: NEWS.rst

.. towncrier release notes start

Twisted 21.7.0.rc1 (2021-07-10)
===============================

Features
--------

> - support py3.10a5 (by fixing twisted.trial.unittest.TestCase.flushWarnings) (#10111)

Reviewer: ```suggestion
- support py3.10b3 (by fixing twisted.trial.unittest.TestCase.flushWarnings) (#10111)
```

> - support py3.10b3 (by fixing twisted.trial.unittest.TestCase.flushWarnings) (#10111)

Reviewer: ```suggestion
- support py3.10b4 (by fixing twisted.trial.unittest.TestCase.flushWarnings) (#10111)
```

Author: Thanks.  I will update.

In practice this release supports  3.10b4...but I think that the ticket number should 10224 and not 10111.

The changes for ticket 10224 will not be part of this release, but since no changes are required between 3.10b3 an 3.10.b4 I think that is safe to advertise that 3.10b4 is supported.

Reviewer: seems good to me

Author: The applied suggestion + the comment from Glyph over the mailing list
```suggestion
- Python 3.10b3 is now supported (#10224)
- Type hinting was added to twisted.internet.defer, making this is the first release
  of Twisted where you might reasonably be able to use mypy without your own custom
  stub files (#10017) 
```

Bugfixes
--------

- twisted.internet.process can now pause and resume producing in python 3 (#9933)

> - When installing Twisted it now requires a minimum Python 3.5.4 version to match the version used with automated testing. This is the minimum Python version that we know that Twisted works with. (#10098)

Reviewer: It's 3.6.7+ now

Reviewer: ```suggestion
- When installing Twisted it now requires a minimum Python 3.6.7 version to match the version used with automated testing. This is the minimum Python version that we know that Twisted works with. (#10098)
```

- Fix type hint for http.Request.uri (from str to bytes). (#10139)
- twisted.web.http_headers.getRawHeaders and twisted.web.http_headers.getAllRawHeaders are now typed to return immutable sequences of header values instead of lists.
  twisted.web.http_headers.getRawHeaders is now typed to return a non-optional value if a non-None default value is given. (#10142)
- Fixed type hint for addr argument to twisted.internet.interfaces.buildProtocol. (#10147)
- twisted.trial._dist.worker.LocalWorker.connectionMade now always writes the
  log file using UTF-8 encoding.
  In previous versions it was using the system default encoding.
  This was causing encoding errors as the distributed trial workers are sending
  Unicode data and the system default encoding might not always be Unicode compatible.
  For example, it can be CP1252 on Windows. (#10157)
- twisted.words.protocols.irc.ctcpExtract was updated to work with PYPY 3.7.4. (#10189)
- twisted.conch.ssh.transport.SSHServerTransport and twisted.conch.ssh.transport.SSHClientTransport no longer use the hardcoded
  SHA1 digest for non-group key exchanges. (#10203)

> - haproxy transport wrapper now returns hosts of type str for getPeer() and getHost(), as specified by IPv4Address and IPv6Address documentation. Previously it was returning bytes for the host.``` (#10211)

Reviewer 2: What's with the backticks here?

Reviewer: ```suggestion
- haproxy transport wrapper now returns hosts of type str for getPeer() and getHost(), as specified by IPv4Address and IPv6Address documentation. Previously it was returning bytes for the host. (#10211)
```

.. towncrier release notes start

Twisted 21.7.0 (2021-07-26)
===========================

Features
--------

- Python 3.10b3 is now supported (#10224)
- Type hinting was added to twisted.internet.defer, making this is the first release
  of Twisted where you might reasonably be able to use mypy without your own custom
  stub files (#10017)

Bugfixes
--------

- The changes to ``DelayedCall.__repr__`` and ``LoopingCall.__repr__`` from
  21.7.0.rc1 were reverted as the wrong assumption that ``__qualname__`` is
  available on all the supported Python versions.
  (#10235)
- The automated release process was updated to generate and release wheel files

>   to PyPy (#10236)

Reviewer: ```suggestion
  to PyPI (#10236)
```

--------

- twisted.internet.defer.inlineCallbacks has an improved type annotation, to avoid typing errors when it is used on a function which returns a non-None result. (#10231)

> - trial.runner.filenameToModule now sets the correct `module.__name__` and `sys.modules` key (#10230)

Reviewer: ```suggestion
- trial.runner.filenameToModule now sets the correct ``module.__name__`` and ``sys.modules`` key (#10230)
```

Reviewer: Double for RST?

Author: I don't know. But I will use the double back-ticks for the final release. Thanks!
